### ISSUE TITLE
False positive expected indentation NaN when disabling ObjectExpressions

### ISSUE THREAD

**Issue by dmfay** [eslint/eslint#9194]
* **ESLint Version:** 4.1.1
* **Node Version:** 8.1.3
* **npm Version:** 5.3.0
**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

In a package.json:

```js
  "eslintConfig": {
    "extends": "eslint:recommended",
    "env": {
      "node": true,
      "mocha": true,
      "es6": true
    },
    "rules": {
      "indent": [
        "error",
        2,
        {
          "ObjectExpression": "off"
        }
      ]
    }
  }
```
</details>

**What did you do? Please include the actual source code causing the issue.**

The above config attempts to disable ESLint indent checks in objects specifically, keeping them for everything else.

```js
function thing(string) {
  const result = where({
    or: [{
      field1: 'value1'
    }, {
      field2: 'value2', field3: 'value3'
    }, {
      field4: 'value4'
    }]
  });
}
```

**What did you expect to happen?**

There should be no error messages regarding indentation in the `where` call.

**What actually happened? Please include the actual, raw output from ESLint.**

```
  5:1   error  Expected indentation of NaN spaces but found 4  indent
  7:1   error  Expected indentation of NaN spaces but found 4  indent
  9:1   error  Expected indentation of NaN spaces but found 4  indent
```

Notably, 4 is the "correct" indentation level without any special rule configuration, so something's going on with toggling `ObjectExpression`s "off". Disabling the `if` in the `ignoreToken` function of lib/rules/indent.js turns the NaN into an actual number, but it's the wrong one and also causes a ton of other false positives so it's not _just_ that.

I tried disabling `ArrayExpression`s and `CallExpression`s as well, to no avail; this is specifically an issue with `ObjectExpression`s.

**Possibly related:**

#8552

**Comment by platinumazure** [comment]
Hi @dmfay, thanks for the issue.

We saw a similar issue with CallExpressions and resolved that in bd09cd5 (released in ESLint 4.5.0), could you try upgrading and see if this is resolved?

**Comment by dmfay** [comment]
That did fix the NaN issue! However, I noticed something else when I checked to make sure that no indent results were raised in the block. Given:

```
function thing(string) {
  const result = where({
      or: [{
        field1: 'value1'
      }, {
      field2: 'value2', field3: 'value3'
    }, {
      field4: 'value4'
    }]
  });
}
```

(lines 3-5, starting with `or`, have been indented an extra level)

The eslint output from 4.5.0 includes:

`  5:1   error  Expected indentation of 4 spaces but found 6  indent`

This seemed like it could be governed by `ArrayExpression` instead of `ObjectExpression` so I turned that off too, but the message still appears.

**Comment by platinumazure** [comment]
I think for objects in arrays, the `indent` rule broadly supports these two patterns:

```js
// pattern 1: Opening brace on same line as opening bracket, closing brace on same line as closing bracket
// Assume 1 level of indentation for key/value pairs and 0 levels for delimiter lines
const foo = [{
  key: "value"
}, {
  key: "value2"
}];

// patter 2: Opening brace on next line after opening bracket, closing brace on previous line from closing bracket
// Assume 2 levels of indentation for key/value pairs and 1 level for delimiter lines
const foo = [
  {
    key: "value"
  },
  {
    key: "value2"
  }
];
```

What I think you might be suggesting is sort of a hybrid style that our `indent` rule does not support:

```js
// Invalid pattern: brace and bracket are on same line but expect extra indentation
const foo = [{
    key: "value"
  }, {
    key: "value2"
}];
```

**Comment by dmfay** [comment]
Yes -- hence trying to disable `ObjectExpression`/`ArrayExpression`, so I can get indent checks elsewhere but ignore them inside complex objects.

**Comment by not-an-aardvark** [comment]
This is working as intended based on https://github.com/eslint/eslint/issues/9194#issuecomment-326638820. To disable checking for this in ESLint 4.5.0+, you should be able to use the `ignoredNodes: ["ObjectExpression"]` option.
